TestController (serviceregistry) is flaky

--- FAIL: TestController (20.00s)
    controller_test.go:113: timeout while waiting (last error: got 0 notifications from controller, want 7)
FAIL
FAIL	istio.io/istio/pilot/pkg/serviceregistry/external	20.110s
=== RUN   TestServices

https://circleci.com/gh/istio/istio/441428?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

--- a/pilot/pkg/serviceregistry/external/controller_test.go
+++ b/pilot/pkg/serviceregistry/external/controller_test.go
@@ -16,7 +16,7 @@ package external_test
 
 import (
 	"fmt"
-	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -34,29 +34,15 @@ func TestController(t *testing.T) {
 	store := memory.Make(configDescriptor)
 	configController := memory.NewController(store)
 
-	countMutex := sync.Mutex{}
-	count := 0
-
-	incrementCount := func() {
-		countMutex.Lock()
-		defer countMutex.Unlock()
-		count++
-	}
-	getCountAndReset := func() int {
-		countMutex.Lock()
-		defer countMutex.Unlock()
-		c := count
-		count = 0
-		return c
-	}
+	count := int64(0)
 
 	ctl := external.NewServiceDiscovery(configController, model.MakeIstioStore(configController))
-	err := ctl.AppendInstanceHandler(func(instance *model.ServiceInstance, event model.Event) { incrementCount() })
+	err := ctl.AppendInstanceHandler(func(instance *model.ServiceInstance, event model.Event) { atomic.AddInt64(&count, 1) })
 	if err != nil {
 		t.Fatalf("AppendInstanceHandler() => %q", err)
 	}
 
-	err = ctl.AppendServiceHandler(func(service *model.Service, event model.Event) { incrementCount() })
+	err = ctl.AppendServiceHandler(func(service *model.Service, event model.Event) { atomic.AddInt64(&count, 1) })
 	if err != nil {
 		t.Fatalf("AppendServiceHandler() => %q", err)
 	}
@@ -97,7 +83,7 @@ func TestController(t *testing.T) {
 			Resolution: networking.ServiceEntry_STATIC,
 		},
 	}
-	expectedCount := 7 // 1 service + 6 instances
+	expectedCount := int64(7) // 1 service + 6 instances
 
 	_, err = configController.Create(cfg)
 	if err != nil {
@@ -105,11 +91,11 @@ func TestController(t *testing.T) {
 	}
 
 	if err := retry.UntilSuccess(func() error {
-		if c := getCountAndReset(); c != expectedCount {
-			return fmt.Errorf("got %d notifications from controller, want %d", c, 7)
+		if gotcount := atomic.AddInt64(&count, 0); gotcount != expectedCount {
+			return fmt.Errorf("got %d notifications from controller, want %d", gotcount, expectedCount)
 		}
 		return nil
-	}, retry.Delay(1*time.Second), retry.Timeout(20*time.Second)); err != nil {
+	}, retry.Delay(50*time.Millisecond), retry.Timeout(5*time.Second)); err != nil {
 		t.Fatal(err)
 	}
 }

